libs/utils/src/postgres_connection.rs

```diff
@@ -0,0 +1,199 @@
+        assert_eq!(host, Host::Domain("hello".to_owned()));
+        assert_eq!(port, None);
+    }
+
+    #[test]
+    fn test_ipv6() {
+        let (host, port) = parse_host_port("[::1]:123").unwrap();
+        assert_eq!(host, Host::<String>::Ipv6(std::net::Ipv6Addr::LOCALHOST));
+        assert_eq!(port, Some(123));
+    }
+
+    #[test]
+    fn test_invalid() {
+        assert!(parse_host_port("hello world").is_err());
+    }
+}
+
+#[derive(Clone)]
+pub struct PgConnectionConfig {
+    host: Host,
+    port: u16,
+    password: Option<String>,
+    options: Vec<String>,
+}
+
+impl PgConnectionConfig {
+    pub fn new_host_port(host: Host, port: u16) -> PgConnectionConfig {
+        PgConnectionConfig {
+            host,
+            port,
+            password: None,
+            options: vec![],
+        }
+    }
+
+    pub fn host(&self) -> &Host {
+        &self.host
+    }
+
+    pub fn port(&self) -> u16 {
+        self.port
+    }
+
+    pub fn set_password(mut self, s: Option<String>) -> Self {
+        self.password = s;
+        self
+    }
+
+    pub fn add_options<I: IntoIterator<Item = S>, S: Into<String>>(mut self, i: I) -> Self {
+        for s in i {
+            self.options.push(s.into());
+        }
+        self
+    }
+
+    /// Return a `<host>:<port>` string.
+    pub fn raw_address(&self) -> String {
+        format!("{}:{}", self.host(), self.port())
+    }
+
+    /// Build a client library-specific connection configuration.
+    /// Used for testing and when we need to add some obscure configuration
+    /// elements at the last moment.
+    pub fn build_tokio_postgres_config(&self) -> tokio_postgres::Config {
+        // Use `tokio_postgres::Config` instead of `postgres::Config` because
+        // the former supports more options to fiddle with later.
+        let mut config = tokio_postgres::Config::new();
+        config.host(&self.host().to_string()).port(self.port);
+        if let Some(password) = &self.password {
+            config.password(password);
+        }
+        if !self.options.is_empty() {
+            config.options(
+                &self
+                    .options
+                    .iter()
+                    .map(|s: &String| s.replace('\\', "\\\\").replace(' ', "\\ "))
+                    .collect::<Vec<_>>()
+                    .join(" "),
+            );
+        }
+        config
+    }
+
+    /// Connect using postgres protocol with TLS disabled.
+    pub fn connect_no_tls(&self) -> Result<postgres::Client, postgres::Error> {
+        postgres::Config::from(self.build_tokio_postgres_config()).connect(postgres::NoTls)
+    }
+}
+
+impl fmt::Debug for PgConnectionConfig {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        struct Redacted {}
+        impl fmt::Debug for Redacted {
+            fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+                f.write_str("REDACTED-STRING")
+            }
+        }
```

Comment:
I don't think this is needed. `Option<&str>` implements Debug.

Reply:
Added a comment. Let me know if you still think it's better to display `"REDACTED"` instead of `REDACTED` to save some lines, your call. That's the way `tokio_postgres::Config` does it and it distinguishes between "a redacted string" and "a string which contains the value `REDACTED`".

Reply:
I see! How about this:

```rust
struct Kek {
    value: u32,
}

impl std::fmt::Debug for Kek {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Kek").field("value", &format_args!("REDACTED")).finish()
    }
}

fn main() {
    println!("{:?}", Kek { value: 10 });
}
```

I find `format_args!` immensely useful, because it's essentially `format!` but w/o allocations.